packages/vite/src/module-runner/moduleCache.ts

```diff
@@ -1,139 +1,110 @@
 import { SOURCEMAPPING_URL } from '../shared/constants'
-import { decodeBase64 } from './utils'
+import { decodeBase64, posixResolve } from './utils'
 import { DecodedMap } from './sourcemap/decoder'
-import type { ModuleCache } from './types'
+import type { ResolvedResult } from './types'
 
 const MODULE_RUNNER_SOURCEMAPPING_REGEXP = new RegExp(
   `//# ${SOURCEMAPPING_URL}=data:application/json;base64,(.+)`,
 )
 
-export class ModuleCacheMap extends Map<string, ModuleCache> {
-  private root: string
-
-  constructor(root: string, entries?: [string, ModuleCache][]) {
-    super(entries)
-    this.root = withTrailingSlash(root)
-  }
-
-  normalize(fsPath: string): string {
-    return normalizeModuleId(fsPath, this.root)
+export class ModuleRunnerNode {
+  public importers = new Set<string>()
+  public imports = new Set<string>()
+  public lastInvalidationTimestamp = 0
+  public evaluated = false
+  public meta: ResolvedResult | undefined
+  public promise: Promise<any> | undefined
+  public exports: any | undefined
+  public file: string
+  public map: DecodedMap | undefined
+
+  constructor(
+    public id: string,
+    public url: string,
+  ) {
+    this.file = cleanUrl(id)
   }
+}
 
-  /**
-   * Assign partial data to the map
-   */
-  update(fsPath: string, mod: ModuleCache): this {
-    fsPath = this.normalize(fsPath)
-    if (!super.has(fsPath)) this.setByModuleId(fsPath, mod)
-    else Object.assign(super.get(fsPath)!, mod)
-    return this
-  }
+export class ModuleRunnerGraph {
+  private root: string
 
-  setByModuleId(modulePath: string, mod: ModuleCache): this {
-    return super.set(modulePath, mod)
-  }
+  public idToModuleMap = new Map<string, ModuleRunnerNode>()
+  public fileToModuleMap = new Map<string, ModuleRunnerNode[]>()
 
-  override set(fsPath: string, mod: ModuleCache): this {
-    return this.setByModuleId(this.normalize(fsPath), mod)
+  constructor(root: string) {
+    this.root = withTrailingSlash(root)
   }
 
-  getByModuleId(modulePath: string): ModuleCache {
-    if (!super.has(modulePath)) this.setByModuleId(modulePath, {})
-
-    const mod = super.get(modulePath)!
-    if (!mod.imports) {
-      Object.assign(mod, {
-        imports: new Set(),
-        importers: new Set(),
-        timestamp: 0,
-      })
-    }
-    return mod
+  public getModuleById(id: string): ModuleRunnerNode | undefined {
+    return this.idToModuleMap.get(id)
   }
 
-  override get(fsPath: string): ModuleCache {
-    return this.getByModuleId(this.normalize(fsPath))
+  public getModulesByFile(file: string): ModuleRunnerNode[] {
+    return this.fileToModuleMap.get(file) || []
   }
 
-  deleteByModuleId(modulePath: string): boolean {
-    return super.delete(modulePath)
+  public getModuleByUrl(url: string): ModuleRunnerNode | undefined {
+    url = unwrapId(url)
+    if (url.startsWith('/')) {
+      const id = posixResolve(this.root, url.slice(1))
+      return this.idToModuleMap.get(id)
+    }
+    return this.idToModuleMap.get(url)
```

Comment:
We probably need to store the `url` map still. For example, this import will load the same module, but `getModuleByUrl` won't return the module for one of them:

```ts
await import('./module')
const moduleName = './module'
await import(moduleName)
```

Reply:
Yes, I think having a `urlToIdModuleMap` would be good.